app.ts

```diff
@@ -0,0 +1,377 @@
+		}
+		catch(e) {
+			console.log('There was a problem with these settings (cannot log in):', e);
+			return false;
+		}
+
+		// @ts-ignore
+		this.homey.settings.set('email', email);
+		// @ts-ignore
+		this.homey.settings.set('password', password);
+		return this.loggedIn;
+	}
+
+	/**
+	 * This function will check if the email or password is either empty, null or undefined, and will
+	 * return a boolean.
+	 */
+	containsAccountSettings() {
+		// @ts-ignore
+		if (!this.homey.settings.get('email')) {
+			return false;
+		}
+		// @ts-ignore
+		if (!this.homey.settings.get('password')) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * This function will make a connection with the Crownstone cloud server and the event server.
+	 */
+	async setupConnections(email: string, password: string) {
+		this.loggedIn = false;
+		try {
+			await this.syncer.login(email, password);
+		}
+		catch(e) {
+			console.log('There was a problem logging into the Crownstone cloud:', e);
+			return;
+		}
+		try {
+			await this.loginToEventServer(email, password);
+		}
+		catch(e) {
+			console.log('There was a problem making a connection with the event server:', e);
+			return;
+		}
+		this.loggedIn = true;
+		console.log('Authenticated with cloud and event servers');
+	}
+
+	/**
+	 * This function will at regular times update the presence information in case we miss some events from the
+	 * event server.
+	 */
+	async pollPresenceData() {
+		this.pollPresenceFunction = setInterval(() => {
+			this.syncer.getPresence();
+		}, 1000 * 60 * this.pollPresenceInterval);
+	}
+
+	/**
+	 * This function will stop the code that listens to SSE events from the SSE server, authenticate and start
+	 * listening again.
+	 */
+	async loginToEventServer(email: string, password: string) {
+		await this.sse.stop();
+		await this.sse.loginHashed(email, password);
+		await this.sse.start(this.sseEventHandler);
+	}
+
+	/**
+	 * The sseEventHandler receives events from the sse-server and fires the runLocationTrigger-function
+	 * when a user enters or leaves a room.
+	 * todo: when the state of a Crownstone changes outside of the app, update the capabilityValue.
+	 *
+	 * Incoming events from the event server are of form (exit event example):
+	 *   {
+	 *     type: 'presence',
+	 *     subType: 'exitLocation',
+	 *     sphere: { id: '$sphereId', name: '$sphereName', uid: 1 },
+	 *     location: { id: '$locationId', name: '$locationName' },
+	 *     user: { id: '$userId', name: '$userName' }
+	 *   }
+	 */
+	sseEventHandler = (data: any) => {
+		//console.log(data);
+		if (data.type === 'system') {
+			// e.g. stream starting
+		}
+		if (data.type === 'presence') {
+			if (data.subType === 'enterLocation') {
+				this.runLocationTrigger(data, true);
+			}
+			if (data.subType === 'exitLocation') {
+				this.runLocationTrigger(data, false);
+			}
+		}
+		// deletion and addition of Crownstones
+		if (data.type === 'dataChange') {
+			if (data.subType === 'stones') {
```

Comment:
Don't we handle the create and trigger a getStones?

Reply:
Done, must be tested